docs/design/apiserver_storage.md

```diff
@@ -0,0 +1,163 @@
+
+## Proposed Storage API
+
+```go
+
+type Storage interface {
+	// Wait waits until the storage has seen any version >= given version.
+	Wait(ctx context.Context, version uint64)
+
+	// Put puts an object with a key based on the given conditions.
+	// See docs on Conditions for more details.
+	Put(ctx context.Context, key string, obj runtime.Object, conditions *Conditions) (cur runtime.Object, err error)
+
+	// Delete a key and its object based on the given conditions.
+	// See docs on Conditions for more details.
+	Delete(ctx context.Context, key string, conditions *Conditions) (cur runtime.Object, err error)
+
+	// Get gets the most recent version of a key.
+	// A key that was modified by this interface will be reflected on read after write succeeded.
+	// If no object exists on the key, it will return not found error.
+	Get(ctx context.Context, key string) (cur runtime.Object, err error)
+
+	// List lists all objects that has given prefix and satisfies selectors.
+	List(ctx context.Context, prefix string, ss ...Selector) (objects []runtime.Object, globalRev uint64, err error)
+
+	// WatchPrefix watches a prefix after given version. If version is 0, we will watch from current state.
+	// It returns notifications of any keys that has given prefix.
+	// Given selectors, it returns events that contained object of interest, either of current and previous.
+	// If there is any problem establishing the watch channel, it will return error.
+	// After channel is established, any error that happened will be returned from WatchChan
+	// immediately before it's closed.
+	WatchPrefix(ctx context.Context, prefix string, version uint64, ss ...Selector) (WatchChan, error)
+
+	// AddIndex adds a new object index.
+	// See docs on FieldValueGetFunc for more details.
+	AddIndex(indexName, field string, g FieldValueGetFunc) error
+	// DeleteIndex deletes an object index.
+	DeleteIndex(indexName string)
+
+	// AddWatcherIndex adds a new watcher index.
+	// See docs on FieldValueGetFunc for more details.
+	AddWatcherIndex(indexName, field string, g FieldValueGetFunc) error
+	// DeleteWatcherIndex deletes a watcher index.
+	DeleteWatcherIndex(indexName string)
+}
+
+
+// Conditions is used in do atomic conditional operations.
+// PrevVersion is used to compare with existing object’s version.
+// Compare would only succeed if versions were matched.
+// Note that PrevVersion=0 means no previous object existed.
+type Conditions struct {
+	PrevVersion unit64
+}
+
+// FieldValueGetFunc returns the value in the object corresponding to given field.
+// field is uniquely indicating a field of a type of object.
+// It returns (value, true) if field exists in the object. Otherwise it returns ("", false).
+type FieldValueGetFunc func(field string, obj runtime.Object) (string, bool)
```

Comment:
Why am I restricted to having a single key for a given object?  In the general case, I  could want to be placed into multiple buckets.  For instance, 

```
type objA struct{
    quotaTheseResources []string
}
```

Someone could then choose to index based on the resources `objA` was quotaing and now we need to have it in multiple buckets.